dbos/_sys_db.py

```diff
@@ -1112,88 +1113,153 @@ def start_queued_workflows(self, queue: "Queue", executor_id: str) -> List[str]:
+                # We will use these to calculate how many more functions this worker can start
+                number_of_tasks_already_started: int = len(
+                    [row[0] for row in rows if row[1] is not None]
+                )
+                dbos_logger.debug(
+                    f"[{queue.name}] {number_of_tasks_already_started} task(s) already started"
                 )
-                .where(SystemSchema.workflow_queue.c.queue_name == queue.name)
-                .where(SystemSchema.workflow_queue.c.completed_at_epoch_ms == None)
-                .order_by(SystemSchema.workflow_queue.c.created_at_epoch_ms.asc())
-            )
-            if queue.concurrency is not None:
-                query = query.limit(queue.concurrency)
 
-            # From the functions retrieved, get the workflow IDs of the functions
-            # that have not yet been started so we can start them.
-            rows = c.execute(query).fetchall()
-            dequeued_ids: List[str] = [row[0] for row in rows if row[1] is None]
-            ret_ids: list[str] = []
-            for id in dequeued_ids:
+                # queue length >= queue.concurrency >= len(rows) >= number_of_tasks_already_started > 0
+                number_of_eligible_tasks: int = (
+                    len(rows) - number_of_tasks_already_started
+                )
+                dbos_logger.debug(
+                    f"[{queue.name}] {number_of_eligible_tasks} task(s) eligible for dequeue"
+                )
+                if number_of_eligible_tasks == 0:
+                    return []
 
-                # If we have a limiter, stop starting functions when the number
-                # of functions started this period exceeds the limit.
-                if queue.limiter is not None:
-                    if len(ret_ids) + num_recent_queries >= queue.limiter["limit"]:
-                        break
+                tasks_this_worker_is_already_working_on: int = len(
+                    [row[0] for row in rows if len(row) == 3 and row[2] == executor_id]
+                )
 
-                # To start a function, first set its status to PENDING and update its executor ID
-                c.execute(
-                    SystemSchema.workflow_status.update()
-                    .where(SystemSchema.workflow_status.c.workflow_uuid == id)
-                    .where(
-                        SystemSchema.workflow_status.c.status
-                        == WorkflowStatusString.ENQUEUED.value
-                    )
-                    .values(
-                        status=WorkflowStatusString.PENDING.value,
-                        executor_id=executor_id,
+                # This worker can dequeue up to whatever is smaller between the eligible tasks and its set concurrency
+                # Of course we must account for tasks this worker is already working on, dequeued during a previous pass of this function
+                max_tasks_this_worker_can_dequeue = int(
```

Comment:
I think there are much simpler ways to express this

Reply:
done